cores/arduino/Adafruit_TinyUSB_Core/Adafruit_USBD_Device.cpp

```diff
@@ -38,6 +38,18 @@
   #define USB_CONFIG_POWER 100
 #endif
 
+// Weak symbols which can be replaced at link time to provide custom
```

Comment:
Thank you for your PR. This is a good approach, but I would prefer a more explicit way to do so with an API, much like the `Adafruit_USBD_Device::setID()` . Let's add more API for this purpose

```
- setManufacturerDescriptor(uint16_t const* desc)
- setProductDescriptor(uint16_t const* desc)
- setLanguageID(uint16_t language_id)
```

default values will be used if these are not set. Note: input value is UTF16 since user may want to use their local language.

If you don't have time, I will eventually implement these when I got my time to work with usb again.

Reply:
Hmm, when would that API run? Aren't the names set and passed to USB in the device constructor at init time, before user code could change it?

The PR is linker logic, the original strings are kicked out by the build process.

Currently the strings are ASCII  and converted internally on-the-fly when called. You want to change that?

Reply:
Those can be called in the setup(),  those aren't set anywhere, `tud_descriptor_string_cb()`  is a callback invoked by tinyUSB stack when receiving GET_DESCRIPTOR request from host. We can literally change how it response anytime (as long as it happens before host ask for it).

The current string is ascii and converted on-the-fly since I know for sure they are all English. If we allow custom descriptor, we should make it possible for user to use their own language.

If you don't have time or not comfortable to do this, don't worry, I will do it later on :)

Reply:
> - setManufacturerDescriptor(uint16_t const* desc)
> - setProductDescriptor(uint16_t const* desc)
> - setLanguageID(uint16_t language_id)

The USB localization is meant to carry multiple languages at the same time in the descriptor. Do you want to store one language id globally in the device for all strings?

> Those can be called in the setup(), ... We can literally change how it response anytime (as long as it happens before host ask for it).

Ah, so we could allow changing it during runtime, just check that the device is not mounted. That sounds nice.

Reply:
Here is a simple version that seem to work nicely:
  https://github.com/kaysievers/ArduinoCore-samd/tree/tinyusb-set-names

It might be really useful to be able to change the name during runtime, as long as the device is not in use by a host.

Reply:
Do you want code to convert utf8 to utf16 properly?

If there isn't anything available, I could copy things from here. I wrote that code many many years ago. :)

  https://git.kernel.org/pub/scm/linux/hotplug/udev.git/tree/src/libudev-util.c#n250

Reply:
> The USB localization is meant to carry multiple languages at the same time in the same descriptor. Do you want to store one language id globally in the device for all strings?

For now, we don't need to support multiple languages, one language is good enough. if user use their local language for manufacturer/product e.g French should set the language ID accordingly to French.

> Ah, so we could allow changing it during runtime, just check that the device is not mounted. That sounds nice.

There is no need to check as well, just set it. It will take affect the next time device is enumerated, user may force an dettach() then attach() for that ( note dettach()/attach() is not yet implemented yet, but soon).

 > Do you want code to convert utf8 to utf16 properly?
> 
> If there isn't anything available, I could copy things from here. I wrote that code many many years ago. :)
> 
> https://git.kernel.org/pub/scm/linux/hotplug/udev.git/tree/src/libudev-util.c#n250

No, the stack doesn't need to convert that, user sketch should give the input as the UTF16 string , which is the native format of string descriptor. 

`setManufacturerDescriptor(uint16_t const* desc)`

ascii --> utf16 is already done with default value, when setting with custom value we should skip the conversion. Sorry if I didn't make it clear enough for you.

Reply:
Updated. It accepts plain UTf-8 now and converts it to simple UTF-16. This way it also accepts internationalized names from the core config, which are 8 bit only.

See the screenshot :)

<img width="770" alt="Screen Shot 2019-08-27 at 11 50 56" src="https://user-images.githubusercontent.com/1064936/63762131-96c1ba80-c8c2-11e9-90b9-dc4ddd1b1f36.png">